### PR TITLE
vw__ sql refactor

### PR DESCRIPTION
**Pull request by inglesp**
This is with an eye to ensuring that derived tables always use current CCG membership, as part of #739.

### CODE REVIEW

**Comment by inglesp** [comment]
As expected the fixture changes in ca592cd cause a bunch of unrelated test failures.  I'm not sure whether it's going to be worth updating these tests, or using a separate set of fixtures for these tests.

**Comment by inglesp** [comment]
@sebbacon This is ready for review.

File: openprescribing/frontend/management/commands/views_sql/chemical_summary_by_ccg.sql [hunk 1]
```diff
@@ -1,14 +1,15 @@
...
-  pct AS pct_id,
+  practices.ccg_id AS pct_id,
   SUBSTR(bnf_code, 1, 9) AS chemical_id,
   SUM(items) AS items,
   SUM(actual_cost) AS cost,
   CAST(SUM(quantity) AS INT64) AS quantity
-FROM
-  {hscic}.normalised_prescribing_standard
+FROM {hscic}.normalised_prescribing_standard
+INNER JOIN {hscic}.practices
```
[the remark this answers is not in the document]
Comment (inglesp) [reply openprescribing/frontend/management/commands/views_sql/chemical_summary_by_ccg.sql]: Aargh aargh aargh sob.

In the tests, data is inserted into `normalised_prescribing_standard` as if it's a table, while of course it's actually a view.

The new test data for this table has one practice change CCG between months, and the tests fail without this corresponding change to the SQL.

However, this wouldn't be possible in production, as the view already joins on `practices` so when querying the view we'll always see a practice in the same CCG.

Please can I have the rest of the day off as compassionate leave?

[the remark this answers is not in the document]
Comment (sebbacon) [reply openprescribing/frontend/management/commands/views_sql/chemical_summary_by_ccg.sql]: See you when you've recovered!

**Comment by inglesp** [comment]
> What is the impact of this bug (changing to only using current practice membership?)

The bug affects analyse queries where the denominator is "total list size" or "STAR PUs" and which aggregate by CCG.  (That is, where the "highlighting" field is set to "a CCG or CCGs".)

At the moment, for a given month the data for the numerator includes all practices currently in the CCG, the denominator includes all practices in the CCG *in that month*.

Once this fix is deployed and `create_views` has been re-run, both the numerator and the denominator will include all practices currently in the CCG.

I don't know how much this will change the data.  We can probably work it out by recalculating `vw__ccgstatistics` (into a new table) and comparing row-by-row.  I'll do this now.

**Comment by inglesp** [comment]
I uploaded the new SQL to /tmp/ccgstatistics_2.sql, and ran the following to insert the CCG statistics into a new table called `vw__ccgstatistics_2`:

```
>>> from gcutils.bigquery import Client
>>> 
>>> client = Client('hscic')
>>> table = client.get_table('vw__ccgstatistics_2')
>>> 
>>> with open('/tmp/ccgstatistics_2.sql') as f:
...     sql = f.read()
... 
>>> substitutions = {'this_month': '2018-02-01'}
>>> 
>>> table.insert_rows_from_query(sql, substitutions=substitutions)
```

I then ran the following query to compare the contents of `vw__ccgstatistics` and `vw__ccgstatistics_2`:

```
SELECT
  t1.date,
  t1.pct_id,
  t1.name,
  t1.total_list_size,
  t2.total_list_size,
  t1.total_list_size / t2.total_list_size AS ratio
FROM
  hscic.vw__ccgstatistics AS t1
INNER JOIN
  hscic.vw__ccgstatistics_2 AS t2
ON
  t1.date = t2.date
  AND t1.pct_id = t2.pct_id
WHERE
  t1.total_list_size != t2.total_list_size
ORDER BY
  ratio
```

This found 775 differing rows (out of 12,568 ) and a CSV of the differences is attached.  I will look at this next week.

[results-20180517-163853.csv.txt](https://github.com/ebmdatalab/openprescribing/files/2013869/results-20180517-163853.csv.txt)

**Comment by inglesp** [comment]
These are notes.  See the next comment for a summary.

Here's a breakdown of the CCGs whose list size changes with the updated SQL, with the number of months with a difference:

```
49 05L NHS SANDWELL AND WEST BIRMINGHAM CCG
49 06W NHS NORWICH CCG
49 02W NHS BRADFORD CITY CCG
49 04X NHS BIRMINGHAM SOUTH AND CENTRAL CCG
49 06Y NHS SOUTH NORFOLK CCG
49 02R NHS BRADFORD DISTRICTS CCG
49 01H NHS NORTH CUMBRIA CCG
49 01K NHS MORECAMBE BAY CCG
45 03G NHS LEEDS SOUTH AND EAST CCG
43 03W NHS EAST LEICESTERSHIRE AND RUTLAND CCG
40 00K NHS HARTLEPOOL AND STOCKTON-ON-TEES CCG
39 02F NHS WEST CHESHIRE CCG
37 13P NHS BIRMINGHAM CROSSCITY CCG
37 08F NHS HAVERING CCG
37 07L NHS BARKING AND DAGENHAM CCG
34 00W NHS CENTRAL MANCHESTER CCG
25 05Y NHS WALSALL CCG
22 01N NHS SOUTH MANCHESTER CCG
 7 00F NHS GATESHEAD CCG
 7 01M NHS NORTH MANCHESTER CCG
 6 00M NHS SOUTH TEES CCG
 3 04D NHS LINCOLNSHIRE WEST CCG
 1 10L NHS ISLE OF WIGHT CCG
```

23 CCGs are affected.  Of these, 4 are closed (the 3 Manchesters, and Gateshead) so can be ignored.

So there are problems with the data for 19 / 213 open CCGs, for analyse queries where the denominator is "total list size" or "STAR PUs" and which aggregate by CCG.

Of these, in most cases the difference in total list size, between the old and new ways of calculating, is small.

Here are the CCGs where the difference is greater than 5% either way.  (I know that averaging an average is a bad idea, but I think we're OK here because the standard deviations are small.)

name | ratio_mean | ratio_std
-- | -- | --
NHS MORECAMBE BAY CCG | 0.449367 | 0.003558
NHS BIRMINGHAM SOUTH AND CENTRAL CCG | 0.864575 | 0.066428
NHS LINCOLNSHIRE WEST CCG | 0.919236 | 0.000000
NHS BRADFORD CITY CCG | 0.942683 | 0.002877
NHS NORWICH CCG | 0.945823 | 0.001102
NHS SOUTH NORFOLK CCG | 1.056124 | 0.001119
NHS NORTH CUMBRIA CCG | 1.601136 | 0.010778

This was generated with the following:

```
df = pd.read_csv('846.csv')
ccgs = df[['code', 'name']].drop_duplicates().set_index('code')
ccgs['ratio_mean'] = df.groupby('code')['ratio'].mean()
ccgs['ratio_std'] = df.groupby('code')['ratio'].std()
ccgs[(ccgs['ratio_mean'] < 0.95) | (ccgs['ratio_mean'] > 1.05)].sort_values('ratio_mean')
```

Morecambe Bay CCG clearly gained a lot of practices between 2017_03 and 2017_06:

```
$ grep -rwc 01K | sort
2015_20/epraccur.csv:24
2016_02/epraccur.csv:24
2016_06/epraccur.csv:24
2016_09/epraccur.csv:26
2016_12/epraccur.csv:26
2017_03/epraccur.csv:26
2017_06/epraccur.csv:64
2017_08/epraccur.csv:65
2017_12/epraccur.csv:65
2018_03/epraccur.csv:65
2018_05/epraccur.csv:65
```

And North Cumbria lost a lot at the same time:

```
$ grep -rwc 01H | sort
2015_20/epraccur.csv:111
2016_02/epraccur.csv:111
2016_06/epraccur.csv:111
2016_09/epraccur.csv:112
2016_12/epraccur.csv:112
2017_03/epraccur.csv:112
2017_06/epraccur.csv:74
2017_08/epraccur.csv:73
2017_12/epraccur.csv:73
2018_03/epraccur.csv:73
2018_05/epraccur.csv:73
```

This was caused be a reorganisation of these two CCGs.  See [here](https://www.cumbriapartnership.nhs.uk/news/new-boundaries-for-ccgs-in-north-lancashire-and-cumbria-are-confirmed-for-2) and #527.

From the [time series charts](https://openprescribing.net/analyse/#org=CCG&orgIds=01K,01H&numIds=6.1.1&denom=total_list_size&selectedTab=chart), it's very obvious that something's up here:

image

For the others, the change in practices is much less dramatic, and looks more like gradual evolution over time.  For example, Birmingham South and Central:

```
$ grep -rwc 04X | sort
2015_20/epraccur.csv:95
2016_02/epraccur.csv:95
2016_06/epraccur.csv:97
2016_09/epraccur.csv:97
2016_12/epraccur.csv:98
2017_03/epraccur.csv:100
2017_06/epraccur.csv:109
2017_08/epraccur.csv:110
2017_12/epraccur.csv:110
2018_03/epraccur.csv:110
2018_05/epraccur.csv:110
```

**Comment by inglesp** [comment]
So to answer:

> What is the impact of this bug (changing to only using current practice membership?)

For 19 currently-open CCGs, we've been showing incorrect historical data in analyse queries where the denominator is "total list size" or "STAR PUs" and which aggregate by CCG.  Specifically, we have calculated the value of the denominator based on historical CCG composition, while we have calculated the value of the numerator based on current CCG composition.

This is particularly serious for Morecambe Bay and North Cumbria, where about half the practices from North Cumbria moved to Morecambe Bay in 2017_04.

image

For 17 other CCGs, the difference in list size is between 86% (Birmingham South and Central) and 106% (South Norfolk).

@sebbacon Can we discuss how to communicate this on the site?

**Comment by sebbacon** [comment]
Thanks. I'm going to put this on the agenda for the next DataLab day.

**Comment by inglesp** [comment]
In the meantime, are you happy for me to merge?

**Pull request merged**